packages/rich-text/src/to-dom.js

```diff
@@ -226,21 +226,17 @@ export function apply( {
 
 export function applyValue( future, current ) {
 	let i = 0;
+	let futureChild;
 
-	while ( future.firstChild ) {
+	while ( ( futureChild = future.firstChild ) ) {
 		const currentChild = current.childNodes[ i ];
-		const futureNodeType = future.firstChild.nodeType;
 
 		if ( ! currentChild ) {
-			current.appendChild( future.firstChild );
-		} else if (
-			futureNodeType !== currentChild.nodeType ||
-			futureNodeType !== TEXT_NODE ||
-			future.firstChild.nodeValue !== currentChild.nodeValue
-		) {
-			current.replaceChild( future.firstChild, currentChild );
+			current.appendChild( futureChild );
+		} else if ( ! currentChild.isEqualNode( futureChild ) ) {
+			current.replaceChild( futureChild, currentChild );
 		} else {
-			future.removeChild( future.firstChild );
+			future.removeChild( futureChild );
```

Comment:
Is `future` used after this function is complete? I'm wondering if there's any gains to be had by avoiding the `removeChild` operation and just stepping through the children.

```js
let i;
for ( i = 0; i < future.childNodes.length; i++ ) {
	const futureChild = future.childNodes[ i ];
	const currentChild = current.childNodes[ i ];

	if ( ! currentChild ) {
		current.appendChild( futureChild );
	} else if ( ! currentChild.isEqualNode( futureChild ) ) {
		current.replaceChild( futureChild, currentChild );
	}
}

while ( current.childNodes[ i ] ) {
	current.removeChild( current.childNodes[ i ] );
}
```

Reply:
This won't work I think. When you append or replace, the node will be gone from `future`. So either we have to remove unused nodes so that `firstChild` is accurate, or we copy all the nods to an array. Not sure if it makes much difference. These nodes are not live.

Reply:
> This won't work I think.

But it passes all tests, so it must work? 😄 

>So either we have to remove unused nodes so that firstChild is accurate, or we copy all the nods to an array.

There might be an option here with iterating backward from the end of `future.childNodes` as well.

>Not sure if it makes much difference. These nodes are not live.

The theory (unconfirmed) is that `removeChild` is a non-trivial operation, even if not live in a document.

Reply:
I think it’s a misconception that the DOM is slow, but I can make test.

Reply:
I put together a benchmark. I'm a bit puzzled by how slow the normal iteration is, but it does show `removeChild` as being sufficiently performant.

Reply:
Yeah the DOM is pretty fast. Just not the live one... :)